src/algebra/pointwise.lean

```diff
@@ -132,8 +132,8 @@ instance [semigroup α] : semigroup (set α) :=
 
 @[to_additive set.add_monoid]
 instance [monoid α] : monoid (set α) :=
-{ mul_one := λ s, by { simp only [← singleton_one, mul_singleton, mul_one, image_id'] },
-  one_mul := λ s, by { simp only [← singleton_one, singleton_mul, one_mul, image_id'] },
+{ mul_one := λ s, by { rw [← singleton_one, mul_singleton], simp only [mul_one, image_id'], },
+  one_mul := λ s, by { rw [← singleton_one, singleton_mul], simp only [one_mul, image_id'], },
   ..set.semigroup,
```

Comment:
Any idea what happened here?

Reply:
I don't understand this properly.

Ordinarily I'd be concerned but in this case I am content to be a little ignorant since the number of changes that I was forced to make as a result of the change the definitions of `monoid`, `add_monoid` are very small and these are very heavily-used types in a very large codebase.

Some remarks for what they're worth as follows. There is a difference in behaviour between `simp only [← singleton_one]` and `rw ← singleton_one`. The former turns the goal `s * 1 = s` into `semigroup.mul s {1} = s` whereas the latter produces `s * {1} = s`. The next lemma `mul_singleton` will only fire (either via `rw` or via `simp`) on the goal in the form `s * {1} = s` so we have to use `rw ← singleton_one`. There is then a similar story for `mul_singleton` itself so I split that off into the `rw` also. However the final two lemmas (`mul_one` and `image_id'`) actually need to be `simp` lemmas. Utterly gross and I do not understand but frankly I'm absolutely out of desire to spend more time on this.

Reply:
We need `simp` for the `mul_one` because it is underneath a binder where `rw` can't reach.

Reply:
Okay, in the interests of getting this PR moving, I've just added a comment in the sources here explaining the reason for needing both `simp only` and `rw`.

I also wish I knew why `simp only` was introducing the `semigroup.mul`. :-(

Reply:
Ah, I found a better fix, adding `mul := (*)`, following the trick Gabriel used in https://github.com/leanprover-community/mathlib/pull/6786/commits/7a913e262fb8c4453204e84fb382b40e8157b3af. Now everything works by `simp only`.